### PR TITLE
Adapt tree-sitter corpus tests for use with SANY

### LINKED ISSUE #882: Work proposal: add corpus of tests for SANY at syntax parsing level
I would like to run this proposal past the maintainers before spending much time on it.

Tree-sitter grammars have a very nice method of writing parser tests, using a collection of .txt files with a very basic DSL; for example:
<details>
  <summary>Sample from tree-sitter-tlaplus test corpus</summary>

```
=============|||
Multiline Conjlist
=============|||

---- MODULE Test ----
op ==
  /\ 
   1
  /\
   2
====

-------------|||

(source_file (module (header_line) (identifier) (header_line)
  (operator_definition (identifier) (def_eq)
    (conj_list
      (conj_item (bullet_conj) (nat_number))
      (conj_item (bullet_conj) (nat_number))
    )
  )
(double_line)))

=============|||
Nested Conjlist
=============|||

---- MODULE Test ----
op ==
 /\ /\ 1
    /\ 2
 /\ 3
====

-------------|||

(source_file (module (header_line) (identifier) (header_line)
  (operator_definition (identifier) (def_eq)
    (conj_list
      (conj_item
        (bullet_conj)
        (conj_list
          (conj_item (bullet_conj) (nat_number))
          (conj_item (bullet_conj) (nat_number))
        )
      )
      (conj_item (bullet_conj) (nat_number))
    )
  )
(double_line)))
```
</details>

Each test involves a title, the source code input, and the expected parse tree output in an S-expression-like DSL.

I would like to take the fairly large corpus of tests already existing in the tree-sitter-tlaplus repo and write a simple translator so they can be run against SANY at the syntax parsing level. This would involve no changes to any SANY code itself. I believe there are the following benefits:

- SANY doesn't currently have any unit tests similar to this, so it seems an unqualified good thing to add them
- sharing test inputs across the various TLA+ parsers (perhaps TLAPM parser could also be added in future) increases incentive to expand the corpus and helps ensure conformance between the parsers
- Using a lightweight DSL such as this instead of encoding the expected parse tree as Java objects greatly reduces the difficulty of adding additional tests

You can see the current tree-sitter-tlaplus test corpus [here](https://github.com/tlaplus-community/tree-sitter-tlaplus/tree/main/test/corpus). There are some tests including unicode symbols which will be omitted.

If this work is performed, will it likely be merged? Follow-up decision if so: it would be nice to upgrade to JUnit 5 (currently we use version 4) so these corpus tests can be emitted at runtime as [dynamic tests](https://junit.org/junit5/docs/5.0.2/api/org/junit/jupiter/api/DynamicTest.html) instead of running all of them inside a single unit test, although this is not strictly necessary.

### PR DESCRIPTION
**Pull request by ahelwer**
Adapts all tree-sitter-tlaplus corpus tests to work with SANY and test it at the syntax processing level. Also greatly expands the number of tests, benefiting both projects.

Turns out translating parse trees from one format to another is a lot of work! Who knew! But I did it, so hurrah for that. Take a look at the monstrous `SanyTranslator.translate` function.

The new unit test class `tla2sany.ParseCorpusTest` round-trips 182 parse inputs and executes in 0.3 seconds.

Closes #882

### CODE REVIEW

File: tlatools/org.lamport.tlatools/.classpath [hunk 1]
```diff
@@ -26,11 +26,15 @@
...
 	<classpathentry kind="lib" path="lib/jline/jline-terminal-3.25.0.jar"/>
-	<classpathentry kind="lib" path="lib/jpf.jar"/>
 	<classpathentry kind="lib" path="lib/jmh/jmh-core-1.21.jar"/>
 	<classpathentry kind="lib" path="lib/jmh/commons-math3-3.2.jar"/>
 	<classpathentry kind="lib" path="lib/gson/gson-2.8.6.jar"/>
 	<classpathentry kind="lib" path="lib/lsp/org.eclipse.lsp4j.debug_0.21.1.v20230829-0012.jar"/>
 	<classpathentry kind="lib" path="lib/lsp/org.eclipse.lsp4j.jsonrpc_0.21.1.v20230829-0012.jar"/>
+	<classpathentry kind="lib" path="lib/junit-4.12.jar"/>
+	<classpathentry kind="lib" path="lib/jpf.jar"/>
+	<classpathentry kind="lib" path="lib/easymock-3.3.1.jar"/>
```
Comment (lemmy) [inline tlatools/org.lamport.tlatools/.classpath:36]: Adding these libraries to .classpath will cause the to be part of the tlatools jar that is part of the Toolbox.

Comment (ahelwer) [reply tlatools/org.lamport.tlatools/.classpath:36]: How then should the project be set up so unit tests can be run from eclipse without having to load in all the toolbox stuff?

Comment (ahelwer) [reply tlatools/org.lamport.tlatools/.classpath]: Reverted these changes in this branch.

**Comment by ahelwer** [comment]
@lemmy this is now ready to review. All these corpus test .txt files are duplicated from the tree-sitter-tlaplus repo, so if you want we could make a third repo called like "unified-syntax-test-corpus" or something to hold them then import it as a submodule. On the other hand it's just a bunch of text files and in-tree duplication can be considered best practice for archival purposes.

File: tlatools/org.lamport.tlatools/test/tla2sany/ParseCorpusTest.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+	 */
+	@Test
+	public void testEntireCorpus() throws ParseException {
+		int testCount = 0;
+		for (CorpusTestFile corpusTestFile : ParseCorpusTest.corpus) {
+			System.out.println(corpusTestFile.path);
+			for (CorpusTest corpusTest : corpusTestFile.tests) {
+				// keeping this here because it is useful to debug why a specific test fails
+				if (!corpusTest.name.equals("Subexpression Tree Navigation")) {
+					//continue;
```
Comment (lemmy) [inline tlatools/org.lamport.tlatools/test/tla2sany/ParseCorpusTest.java:64]: ?

File: tlatools/org.lamport.tlatools/test/tla2sany/SanyTranslator.java [hunk 3]
```diff
@@ -0,0 +1,1633 @@
...
+
+/**
+ * Handles the translation of raw SANY output to the test file AST DSL.
+ */
+public class SanyTranslator {	
+
+	/**
+	 * Sometimes SANY provides very little structure in its parse output,
+	 * basically just a flat list of tokens. This is most common when dealing
+	 * with constructs that SANY doesn't care much about, like proof-related
```
Comment (lemmy) [inline tlatools/org.lamport.tlatools/test/tla2sany/SanyTranslator.java:27]: How much is missing in SANY to add more structure for proof-related syntax?  There is an unfinished attempt to replace TLAPS' parser with SANY and this sounds like a potential show-stopper for this effort.

File: tlatools/org.lamport.tlatools/test/tla2sany/ParseCorpusTest.java (hunk 2 above)
Comment (ahelwer) [reply tlatools/org.lamport.tlatools/test/tla2sany/ParseCorpusTest.java:64]: For future development and adding of tests. The main way I translated SANY's parse tree was running the debugger and seeing what SANY emits at a given point in the giant switch statement. Some of the rules are run through hundreds of times if you run the entire corpus, so it's useful to be able to pick out a single test to run so you get the debugger state in a specific rule for that test. That's what this code stub does, if you uncomment continue. It can be removed but it might be useful to the next person (probably me) who adds tests.

File: tlatools/org.lamport.tlatools/test/tla2sany/SanyTranslator.java (hunk 3 above)
Comment (ahelwer) [reply tlatools/org.lamport.tlatools/test/tla2sany/SanyTranslator.java:27]: Not really that much. I would say use-or-hide is the main culprit, which you can see because I had to break that parsing logic out into its own set of functions instead of writing it in the main switch statement. It's annoying but surmountable. The parse output of SANY could also be improved.

**Comment by lemmy** [comment]
> @lemmy this is now ready to review. All these corpus test .txt files are duplicated from the tree-sitter-tlaplus repo, so if you want we could make a third repo called like "unified-syntax-test-corpus" or something to hold them then import it as a submodule. On the other hand it's just a bunch of text files and in-tree duplication can be considered best practice for archival purposes.
> 

Redundancy doesn't hurt.  Redundancy doesn't hurt.  However, I'd suggest to add a big fat warning here and in the tree-sitter-tlaplus repo to always modify the files in both places.

File: tlatools/org.lamport.tlatools/test/tla2sany/SanyTranslator.java (hunk 3 above)
Comment (lemmy) [reply tlatools/org.lamport.tlatools/test/tla2sany/SanyTranslator.java:27]: Makes sense, I suggest adding this note to the code.

File: tlatools/org.lamport.tlatools/test/tla2sany/ParseCorpusTest.java (hunk 2 above)
Comment (ahelwer) [reply tlatools/org.lamport.tlatools/test/tla2sany/ParseCorpusTest.java:68]: Added this explanation as a comment in the code.

File: tlatools/org.lamport.tlatools/test/tla2sany/SanyTranslator.java (hunk 3 above)
Comment (ahelwer) [reply tlatools/org.lamport.tlatools/test/tla2sany/SanyTranslator.java:42]: Added this note as a comment in the code here.

**Comment by Calvin-L** [comment]
> (@Calvin-L are you going to review it as well?)

I skimmed over it, but I'm not planning to do a deep review. I'm happy to see this change, and the two things that matter most to me are
1. This is test-only change, so it's essentially zero-risk.
2. It has already proved its worth by discovering lots of corner cases like #884, #885, and #888. :)
